### ISSUE TITLE
"Refreshing Blitz Cache" Hogging Queue; "Checking Element Queries"

### ISSUE THREAD

**Issue by ryanmasuga** [putyourlightson/craft-blitz#145]
### Question
We have a site where we do a lot of importing of content (in this case, from Airtable) using cron jobs. The front end of the site is largely "brochure-like" in that there are no logged in/out pages or anything really complex so we thought we could blitz the whole site in a straightforward way. And it's true - the site is fast, and Blitz works well. 

The only issue we're having it blitz and the queue. The client contacted us, concerned about the queue in the backend, because the backed up queue seems to be hindering CP page loads. The queue never finishes on this site - it's always many 100's of lines ling. 

### Additional context
Here's  a screenshot of the very top of our queue:

<img width="463" alt="blitz" src="https://user-images.githubusercontent.com/988995/70580009-51c74880-1b80-11ea-8afb-ba5aeadabaed.png">

There are 100's of blitz lines, followed by countless "Updating Search Indexes" - that it seems the queue will never get to, behind all the blitz checking that needs to be done!

We're running the queue from the CLI with a cron (`runQueueAutomatically' => false` in general.php), and we also have the Queue Manager plugin installed, as well as AsyncQueue. 

This site has lots of available memory (I have it set to 2048M) and setting max_execution_time is not an issue (it's 300 right now) - should we even need to keep upping that.

Concurrency = 3
Warm Cache Automatically = false
Clear Cache Automatically = true

I'm thinking maybe the Clear Cache Automatically setting should be false, because we're constantly updating or adding entries via imports, and I'm guessing due to the number of relationships on the front end, any time an entry gets updated, all those element queries need to get checked? Perhaps we should only clear cache automatically every so often via cron job?

Perhaps there's another more specific suggestion or two you have? Happy to give you any sort of details or access you need, Ben! Thanks.

**Comment by bencroker** [comment]
Thanks for the detailed report Ryan!

For a "brochure-like" site, it certainly contains a lot of element queries (close to 20,000) and that is part of the reason for the long-running queue jobs. The cause of this is likely to be a lack of use of eager-loading elements in the templates and complex relationships between elements.

If content is constantly being updated via imports, you probably should disable the [Clear Cache Automatically](https://putyourlightson.com/plugins/blitz#clear-cache-automatically) setting and instead set up a `refresh-expired` [cron job](https://putyourlightson.com/plugins/blitz#cron-jobs) at appropriate intervals. The _Refresh Blitz cache_ job will still have to run though in order to mark pages as expired, so it won't solve everything.

The best approach depends on the nature of the site. How often is content being updated and how timely should it be reflected on the site? How big is the site and does it contain mostly detail pages or listing pages as well? Is the site live and could you share it (either here or privately)?

One approach you could take to eliminate long-running jobs completely, is to set `'cacheElementQueries' => false,` in the config settings and refresh the cache using a [console command](https://putyourlightson.com/plugins/blitz#console-commands) with a cron job at a regular interval (for example once an hour).

**Comment by ryanmasuga** [comment]
Thanks for the feedback Ben. Here's what we did for now:

- Stopped all the currently running Blitz refreshing
- Flushed and cleared the Blitz cache entirely
- Set `'clearCacheAutomatically' => false,`
- Set `'cacheElementQueries' => false,`
- Set up a cron job to run `blitz/cache/refresh-expired` every 30 minutes

Within minutes of making these changes there were 600+ cached pages again (pretty decent traffic on this site), and the site working really well in the past 30-40 minutes so far. I'm keeping an eye on it, and can adjust from here.

Sounds like`blitz/cache/refresh-expired` wouldn't address caches that have `NULL` for an expiryDate. Now that we're not automatically refreshing the cache and we're not caching the Element Queries, If we wanted to ensure every page had some sort of expiryDate we will either need to set a new default (other than 0) for `cacheDuration` or  get into [Page Specific Options](https://putyourlightson.com/plugins/blitz#page-specific-options) where we can set more specific durations. Does that sound correct? Or even add tags and run a console command to clear specific tags on different schedules, based on how "critical" the content in the tag is, like your example here:

```
// Refresh cache tagged with the tags “home” and “listing” every day at 6am.
0 6 * * * /usr/bin/php /path/to/craft blitz/cache/refresh-tagged home,listing
```

**Comment by bencroker** [comment]
Ok so it looks like you took all of what I said and applied, but it won't help much in that configuration. If you can let me know a bit about the site in question (including my previous questions) then I can suggest a configuration that I think will work best :)

**Comment by bencroker** [comment]
With the `cacheElementQueries` setting disabled, listing pages will not be refreshed whenever a _new_ entry should appear on them (including paginated lists). Everything else will refresh correctly. So what I propose is that you add one or more tags using [page specific options](https://putyourlightson.com/plugins/blitz#page-specific-options) to every listing page and run a [cron job](https://putyourlightson.com/plugins/blitz#cron-jobs) at a regular interval to refresh those pages with the command `./craft blitz/cache/refresh-tagged`.

So to summarise:

1. Set `clearCacheAutomatically` back to `true` so cache clearing will work.
2. Set `cacheElementQueries` to `false` to avoid long-running jobs.
2. Remove the `blitz/cache/refresh-expired` cron job (unless you have another reason to use it).
4. Add tags to all of the listing pages.
5. Add a `blitz/cache/refresh-tagged` cron job passing in the tag(s) to refresh at regular intervals.

**Comment by bencroker** [comment]
Any other questions or can this be closed?

**Comment by ryanmasuga** [comment]
This is all set. Thanks for all your help and the clear answers and suggestions!

**Comment by bencroker** [comment]
Excellent, glad to hear it!
